dot/rpc/websocketListeners.go

```diff
@@ -0,0 +1,368 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+package rpc
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/ChainSafe/gossamer/dot/rpc/modules"
+	"github.com/ChainSafe/gossamer/dot/state"
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// Listener interface for functions that define Listener related functions
+type Listener interface {
+	Listen()
+}
+
+func (c *WSConn) startListener(lid int) {
+	go c.subscriptions[lid].Listen()
+}
+
+func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (int, error) {
+	if c.storageAPI == nil {
+		c.safeSendError(reqID, nil, "error StorageAPI not set")
+		return 0, fmt.Errorf("error StorageAPI not set")
+	}
+
+	scl := &StorageChangeListener{
+		channel: make(chan *state.SubscriptionResult),
+		wsconn:  c,
+	}
+	sub := &state.StorageSubscription{
+		Filter:   make(map[string][]byte),
+		Listener: scl.channel,
+	}
+
+	pA, ok := params.([]interface{})
+	if !ok {
+		return 0, fmt.Errorf("unknow parameter type")
+	}
+	for _, param := range pA {
+		switch p := param.(type) {
+		case []interface{}:
+			for _, pp := range param.([]interface{}) {
+				data, ok := pp.(string)
+				if !ok {
+					return 0, fmt.Errorf("unknow parameter type")
+				}
+				sub.Filter[data] = []byte{}
+			}
+		case string:
+			sub.Filter[p] = []byte{}
+		default:
+			return 0, fmt.Errorf("unknow parameter type")
+		}
+	}
+
+	chanID, err := c.storageAPI.RegisterStorageChangeChannel(*sub)
+	if err != nil {
+		return 0, err
+	}
+	scl.chanID = chanID
+
+	c.qtyListeners++
+	scl.subID = c.qtyListeners
+	c.subscriptions[scl.subID] = scl
+	c.storageSubChannels[scl.subID] = chanID
+
+	initRes := newSubscriptionResponseJSON(scl.subID, reqID)
+	c.safeSend(initRes)
+
+	return scl.subID, nil
+}
+
+// StorageChangeListener for listening to state change channels
+type StorageChangeListener struct {
+	channel chan *state.SubscriptionResult
+	wsconn  *WSConn
+	chanID  byte
+	subID   int
+}
+
+// Listen implementation of Listen interface to listen for importedChan changes
+func (l *StorageChangeListener) Listen() {
+	for change := range l.channel {
+		if change == nil {
+			continue
+		}
+
+		result := make(map[string]interface{})
+		result["block"] = change.Hash.String()
+		changes := make([][]string, len(change.Changes))
+		for _, v := range change.Changes {
+			kv := []string{common.BytesToHex(v.Key), common.BytesToHex(v.Value)}
+			changes = append(changes, kv)
```

Comment:
`append` adds an element to the end of the array. `changes` slice will be 2x size.

```
func main() {
	const arrSize = 5

	// Incorrect code: Final length will be 10.
	vals := make([]int, arrSize) // Creates [0,0,0,0,0]
	for i := 0; i < 5; i++ {
		vals = append(vals, i)
	}
	fmt.Println(len(vals))
	fmt.Println(vals)

	// Output:
	// 10
	// [0 0 0 0 0 0 1 2 3 4]

	// Solution 1 (Preferred): Pre declare length = 0 and capacity = 5. Creates []
	vals = make([]int,0, arrSize)
	for i := 0; i < 5; i++ {
		vals = append(vals, i)
	}
	fmt.Println(len(vals))
	fmt.Println(vals)

	// Output:
	// 5
	// [0 1 2 3 4]

	// Solution 2:
	vals = make([]int, arrSize) // Creates [0,0,0,0,0]
	for i := 0; i < 5; i++ {
		vals[i] = i
	}
	fmt.Println(len(vals))
	fmt.Println(vals)
	// Output:
	// 5
	// [0 1 2 3 4]
}
```